crates/bevy_ecs/src/query/access.rs

```diff
@@ -773,38 +774,97 @@ impl<T: SparseSetIndex> Access<T> {
-        (
-            self.component_read_and_writes
-                .ones()
-                .map(T::get_sparse_set_index),
-            self.component_read_and_writes_inverted,
-        )
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use bevy_ecs::query::{Access, ComponentAccessKind};
+    /// let mut access = Access::<usize>::default();
+    ///
+    /// access.add_component_read(1);
+    /// access.add_component_write(2);
+    /// access.add_archetypal(3);
+    ///
+    /// let result = access
+    ///     .try_iter_component_access()
+    ///     .map(Iterator::collect::<Vec<_>>);
+    ///
+    /// assert_eq!(
+    ///     result,
+    ///     Ok(vec![
+    ///         ComponentAccessKind::Shared(1),
+    ///         ComponentAccessKind::Exclusive(2),
+    ///         ComponentAccessKind::Archetypal(3),
+    ///     ]),
+    /// );
+    /// ```
+    pub fn try_iter_component_access(
+        &self,
+    ) -> Result<impl Iterator<Item = ComponentAccessKind<T>> + '_, UnboundedAccess> {
+        if self.component_writes_inverted || self.component_read_and_writes_inverted {
+            return Err(UnboundedAccess {
+                writes_inverted: self.component_writes_inverted,
+                read_and_writes_inverted: self.component_read_and_writes_inverted,
+            });
+        }
+
+        let reads_and_writes = self.component_read_and_writes.ones().map(|index| {
+            let sparse_index = T::get_sparse_set_index(index);
+
+            if self.component_writes.contains(index) {
+                ComponentAccessKind::Exclusive(sparse_index)
+            } else {
+                ComponentAccessKind::Shared(sparse_index)
+            }
+        });
+
+        let archetypal = self
+            .archetypal
+            .ones()
+            .filter(|&index| {
+                !self.component_writes.contains(index)
+                    && !self.component_read_and_writes.contains(index)
+            })
+            .map(|index| ComponentAccessKind::Archetypal(T::get_sparse_set_index(index)));
+
+        Ok(reads_and_writes.chain(archetypal))
     }
+}
 
-    /// Returns an iterator over the component IDs that this `Access` either
-    /// writes or can't write.
-    ///
-    /// The returned flag specifies whether the list consists of the components
-    /// that the access *can* write (false) or whether the list consists of the
-    /// components that the access *can't* write (true).
-    pub(crate) fn component_writes(&self) -> (impl Iterator<Item = T> + '_, bool) {
-        (
-            self.component_writes.ones().map(T::get_sparse_set_index),
-            self.component_writes_inverted,
-        )
+/// Error returned when attempting to iterate over items included in an [`Access`]
+/// if the access excludes items rather than including them.
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Error)]
+#[error("Access is unbounded")]
+pub struct UnboundedAccess {
```

Comment:
Do we postfix our errors with `Error`?

```suggestion
pub struct UnboundedAccessError {
```